tensorboard/data/server/types.rs

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Core simple types.
+
+use std::borrow::Borrow;
+
+/// A step associated with a record, strictly increasing over time within a record stream.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Copy, Clone)]
+pub struct Step(pub i64);
+
+/// The wall time of a TensorBoard event.
+///
+/// Wall times represent floating-point seconds since Unix epoch. They must be finite and non-NaN.
+#[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
+pub struct WallTime(f64);
+
+impl WallTime {
+    /// Parses a wall time from a time stamp representing seconds since Unix epoch.
+    ///
+    /// Returns `None` if the given time is infinite or NaN.
+    pub fn new(time: f64) -> Option<Self> {
+        if time.is_finite() {
+            Some(WallTime(time))
+        } else {
+            None
+        }
+    }
+}
+
+// Wall times are totally ordered and have a total equivalence relation, since we guarantee that
+// they are not NaN.
+#[allow(clippy::derive_ord_xor_partial_ord)] // okay because it agrees with `PartialOrd` impl
+impl Ord for WallTime {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.partial_cmp(&other)
+            .unwrap_or_else(|| unreachable!("{:?} <> {:?}", &self, &other))
+    }
+}
+impl Eq for WallTime {}
```

Comment:
It does the same thing that `derive(Eq)` would do, since `Eq` itself has
no methods (it’s a marker interface). But you’re only permitted to
`derive(Eq)` on types all of whose fields are also `Eq`. In this case,
we have a field of type `f64`, which is _not_ `Eq` (because of NaN), so
we cannot use `derive(Eq)`:

```diff
$ git diff -U1
diff --git a/tensorboard/data/server/types.rs b/tensorboard/data/server/types.rs
index 59a764504..755329912 100644
--- a/tensorboard/data/server/types.rs
+++ b/tensorboard/data/server/types.rs
@@ -26,3 +26,3 @@ pub struct Step(pub i64);
 /// Wall times represent floating-point seconds since Unix epoch. They must be finite and non-NaN.
-#[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
+#[derive(Debug, PartialEq, PartialOrd, Copy, Clone, Eq)]
 pub struct WallTime(f64);
@@ -51,3 +51,2 @@ impl Ord for WallTime {
 }
-impl Eq for WallTime {}
 
```

```
$ cargo check
error[E0277]: the trait bound `f64: std::cmp::Eq` is not satisfied
   --> types.rs:28:21
    |
28  | pub struct WallTime(f64);
    |                     ^^^ the trait `std::cmp::Eq` is not implemented for `f64`
    | 
```

Instead, `impl Eq for WallTime {}` is our explicit way of declaring that
our type really does have a total equivalence relation, even though the
types of its fields do not.